_posts/2021-01-11-zappa-serverless-docker.md

```diff
@@ -0,0 +1,224 @@
+---
+layout: post
+title: Serverless Docker on AWS Lambda with Zappa
+author: ianwhitestone
+summary: Announcing Zappa's new container image support
+comments: true
+image: images/zappa-serverless-docker/main.png
+---
+
+{% include head.html %}
+
+<p align="center">
+    <img width="70%" src="{{ site.baseurl }}{% link images/zappa-serverless-docker/main.png %}">
+</p>
+
+I've been using [Zappa](https://github.com/Miserlou/Zappa) to manage my serverless Python deployments for over a year now. Zappa is magic. If you've ever tried to deploy an [Amazon Web Services (AWS) Lambda](https://aws.amazon.com/lambda/) function from scratch, you know the pains. Creating the required Identity and Access Management (IAM) roles, building a zip file with the function and any dependencies, ensuring you have the proper file permissions in that zip, adding the required WSGI middleware, testing your function with mock data, retrieving the output logs, linking your function to [API gateway](https://aws.amazon.com/api-gateway/), and the list goes on. Zappa takes care of all of this. 
+
+But even with all its magic, there are still some major pains for many Python deployments. Your Lambda zip package can only be a [few hundred megabytes](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html), which is easily exceeded as soon as you have any larger dependencies like pandas, numpy, or sklearn. Zappa provides a smart [workaround](https://github.com/Miserlou/Zappa#large-projects) for this by saving your large dependencies in S3 and downloading them at runtime, but this kills response times when your function is [cold starting](https://www.serverless.com/blog/keep-your-lambdas-warm). On top of the size limits, you also need to make sure you are using Python package versions that were compiled for Amazon Linux for any packages with C dependencies. These two nuances quickly add more complexity & headaches to your serverless deployments. 
+
+Last month, [AWS announced new functionality](https://aws.amazon.com/blogs/aws/new-for-aws-lambda-container-image-support/) that lets you deploy Lambda functions as **container images** up to 10GB in size. What does that mean? Instead of pre-compiling all your packages on Amazon Linux, putting them & your application code in a zip file and hoping you are under the limit, you can now just create a Docker image and deploy your function with that. 
```

Comment:
With this approach, the challenges I outlined above are no longer a concern.

Reply:
```suggestion
Last month, [AWS announced new functionality](https://aws.amazon.com/blogs/aws/new-for-aws-lambda-container-image-support/) that lets you deploy Lambda functions as **container images** up to 10GB in size. What does that mean? Instead of pre-compiling all your packages on Amazon Linux, putting them & your application code in a zip file and hoping you are under the limit, you can now just create a Docker image and deploy your function with that. With this approach, the challenges I outlined above are no longer a concern.
```